packages/@aws-cdk/aws-codebuild/README.md

```diff
@@ -109,6 +109,38 @@ aws codebuild import-source-credentials --server-type GITHUB --auth-type PERSONA
 
 This source type can be used to build code from a BitBucket repository.
 
+## Caching
+
+You can save time when your project builds by using a cache. A cache can store reusable pieces of your build environment and use them across multiple builds. Your build project can use one of two types of caching: Amazon S3 or local. In general, S3 caching is a good option for small and intermediate build artifacts that are more expensive to build than to download. Local caching is a good option for large intermediate build artifacts because the cache is immediately available on the build host.
+
+### S3 Caching
+
+With S3 caching, the cache is stored in an S3 bucket which is available from multiple hosts.
+
+```typescript
+new codebuild.Project(stack, 'Project', {
+  source: new codebuild.CodePipelineSource(),
+  cacheBucket: new Bucket(stack, 'Bucket')
+});
+```
+
+### Local Caching
+
+With local caching, the cache is stored on the codebuild instance itself. Three different cache modes are supported:
```

Comment:
One thing that's not clear from this is what causes a reuse of an instance? When is the local cache being utilized.

Reply:
Any reason why this we shouldn't enable this by default?

@skinny85

Reply:
This also isn't entirely clear to me to be honest. The only hint regarding this is that "_After a build, AWS CodeBuild maintains the cache for some time in anticipation of another build._" (https://aws.amazon.com/about-aws/whats-new/2019/02/aws-codebuild-now-supports-local-caching/). My documentation is largely copy/pasted from the [official documentation](https://docs.aws.amazon.com/codebuild/latest/userguide/build-caching.html) where this also isn't clearly mentioned.

Reply:
@Kaixiang-AWS can you weigh in here? Thanks!

Reply:
CodeBuild cannot guarantee a reuse of instance. It's a best effort. Imagine you have a build ran and cached on an instance, and then you have 2 builds start at the same time. Only one of them will utilize the cache and the other one will have to run on a new instance which doesn't have cache in it.
Anyway, I think the problem here is we might miss a modeling opportunity if we do it this way. Defining a base cache class and extending it to S3 or local caching might be a more feasible and extensible way. If there are more cache types in the future, in this way we only need to add another class which extends base class and don't need to handle the complicated input validation.

Reply:
Thanks! Can we put this info into the README?

Reply:
Done! I also looked into the refactor that @Kaixiang-AWS mentioned. I'm currently not experienced enough with Typescript and/or the CDK to properly implement this. Can we for now merge this PR and create a new issue to refactor it? The feature would at least become available if we merge the PR.

Reply:
I like @Kaixiang-AWS's suggestion. It's a pattern we've been using we call "union-like classes" and it fits this API perfectly. I rather we spend a bit more time now than introduce a breaking change later.

Here's a sketch:

```ts

export interface BucketCacheOptions {
  cacheDir?: string;
}

export enum LocalCacheMode {
  Source = 'LOCAL_SOURCE_CACHE',
	DockerLayer = 'LOCAL_DOCKER_LAYER_CACHE',
	Custom = 'LOCAL_CUSTOM_CACHE',
}

export abstract class Cache {
  public static none(): Cache {
    return { _toCloudFormation: () => undefined };
  }

  public static bucket(bucket: s3.IBucket, options: BucketCacheOptions): Cache {
    const cacheDir = options.cacheDir ? options.cacheDir : Aws.noValue;
    return { _toCloudFormation: () => ({
      type: 'S3',
      location: Fn.join('/', [ bucket.bucketName, cacheDir ])})
    };
  }

  public static local(...modes: LocalCacheMode[]): Cache {
    return {
      _toCloudFormation: () => ({
        type: 'LOCAL',
        modes
      })
    };
  }

  /**  @internal */
  public abstract _toCloudFormation(): codebuild.CfnProject.ProjectCacheProperty | undefined;
}

export interface CommonProjectProps {
  /**
   * @default Cache.none
   */
  readonly cache?: Cache;
}
```

Reply:
Looks good, I'll dive into it 👍